## 8) Com relação à aritmética inteira responda:

In [6]:
    function b=SomaBinaria(b1,b2)
    // Soma de números binários  b=b1+b2
    if (length(b1)<>length(b2)) then
        disp("b1 e b2 devem ter o mesmo número de bits")
        return
    end
    n=length(b1)
    v1=0
    for(k=n:-1:1)
        b(1,k)=b1(k)+b2(k)+v1
        v1=0
        if b(k)==2 then
            b(k)=0
            v1=1
        elseif b(k)==3 then
            b(k)=1
            v1=1   
        end
    end
endfunction

In [7]:
function a=binario2dec3(b_c2,n)
   // b INTEIRO POSITIVO OU NEGATIVO
   // com COMPLEMENTO 2  
    b=b_c2
    if(b(1)==1) then   //complemento 2  
        b=bitcmp(b_c2,1)   // inverter bits
        b=SomaBinaria(b,[zeros(1:n-1) 1]) // somar 1
    end
    p= 2^(n-1:-1:0);
    a= (-1)^b_c2(1) * (b*p');  
endfunction


In [8]:
function b_c2 = dec2binario3(a,n) 
    // Conversão de decimal INTEIRO POSITIVO OU NEGATIVO para binário
    // com COMPLEMENTO 2  -  Inverter bits e somar 1
  b=[]
  if(  a<(-2^(n-1)) |  a>(2^(n-1)-1)  ) then
      printf("%d não pode ser reprentado com %d bits",a,n)
      return
  end
  p= 2^(n-1:-1:0);
  r=int(abs(a)./p)
  b = modulo(r,2) //conversão do decimal 'a' para o binário 'b' com n bits
  b_c2=b
  if(a<0) then   //complemento 2  
      b_c2=bitcmp(b,1)   // inverter bits
      b_c2=SomaBinaria(b_c2,[zeros(1:n-1) 1])  // somar 1
  end
  s=""
  for i=1:n
    s=s+string(b_c2(i))
  end
  printf("%f (B10) = %s (B2)\n",a,s)
  a2=binario2dec3(b_c2) // conversão do binário b_ce para decimal a2
  printf("%s (B2) = %f (B10)\n",s,a2)
endfunction

---
### A) Se usarmos uma máquina calculadora, que utiliza inteiros na base 2 com 8 bits e sinal em complemento 2, para subtrairmos o número decimal 43 do número binário 01110011, qual será o resultado em binário? e em decimal?

#### R: base2   = 10111000
#### R: base10 = -72

In [45]:
bits = 8;
dec_a = 43;
bin_b = [0, 1, 1, 1, 0, 0, 1, 1];

In [46]:
dec_b = binario2dec3(bin_b, bits)

  dec_b  = 

   115.



In [39]:
dec_answer = dec_a - dec_b

  dec_answer  = 

  -72.



In [47]:
dec2binario3(dec_answer, bits)

 -37.000000 (B10) = 11011011 (B2)
11011011 (B2) = -37.000000 (B10)
 ans  =

   1.   1.   0.   1.   1.   0.   1.   1.



---
### B) Se usarmos uma máquina calculadora, que utiliza inteiros na base 2 com 8 bits e sinal em complemento 2, para subtrairmos o número decimal 41 do número binário 01001110, qual será o resultado em binário? e em decimal?

#### R: base2   = 11011011
#### R: base10 = -37

In [49]:
dec_a = 41;
bin_b = [0,1,0,0,1,1,1,0];

In [50]:
dec_b = binario2dec3(bin_b, bits)

  dec_b  = 

   78.



In [51]:
dec_answer = dec_a - dec_b

  dec_answer  = 

  -37.



In [52]:
dec2binario3(dec_answer, bits)

 -37.000000 (B10) = 11011011 (B2)
11011011 (B2) = -37.000000 (B10)
 ans  =

   1.   1.   0.   1.   1.   0.   1.   1.



---
### C) Se usarmos uma máquina calculadora, que utiliza inteiros na base 2 com 8 bits e sinal em complemento 2, para subtrairmos o número decimal 33 do número binário 01010110, qual será o resultado em binário?

#### R: base2   = 11001011
#### R: base10 = -53

In [54]:
dec_a = 33;
bin_b = [0,1,0,1,0,1,1,0];

In [55]:
dec_b = binario2dec3(bin_b, bits)

  dec_b  = 

   86.



In [56]:
dec_answer = dec_a - dec_b

  dec_answer  = 

  -53.



In [57]:
dec2binario3(dec_answer, bits)

 -53.000000 (B10) = 11001011 (B2)
11001011 (B2) = -53.000000 (B10)
 ans  =

   1.   1.   0.   0.   1.   0.   1.   1.

